csclient/csclient.go

```diff
@@ -0,0 +1,487 @@
+	// Validate the entity id.
+	if id.Series == "" {
+		return nil, errgo.Newf("no series specified in %q", id)
+	}
+	if id.Revision != -1 {
+		return nil, errgo.Newf("revision specified in %q, but should not be specified", id)
+	}
+
+	// Prepare the request.
+	req, err := http.NewRequest("POST", "", nil)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot make new request")
+	}
+	req.Header.Set("Content-Type", "application/zip")
+	req.ContentLength = size
+
+	// Send the request.
+	resp, err := c.DoWithBody(req, "/"+id.Path()+"/archive?hash="+hash, httpbakery.SeekerBody(body))
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot post archive")
+	}
+	defer resp.Body.Close()
+
+	// Parse the response.
+	var result params.ArchiveUploadResponse
+	if err := parseResponseBody(resp.Body, &result); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return result.Id, nil
+}
+
+// PutExtraInfo puts extra-info data for the given id.
+// Each entry in the info map causes a value in extra-info with
+// that key to be set to the associated value.
+// Entries not set in the map will be unchanged.
+func (c *Client) PutExtraInfo(id *charm.Reference, info map[string]interface{}) error {
+	req, _ := http.NewRequest("PUT", "", nil)
+	req.Header.Set("Content-Type", "application/json")
+	data, err := json.Marshal(info)
+	if err != nil {
+		return errgo.Notef(err, "cannot marshal extra-info")
+	}
+	body := bytes.NewReader(data)
+	resp, err := c.DoWithBody(req, "/"+id.Path()+"/meta/extra-info", httpbakery.SeekerBody(body))
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	resp.Body.Close()
+	return nil
+}
+
+// Meta fetches metadata on the charm or bundle with the
+// given id. The result value provides a value
+// to be filled in with the result, which must be
+// a pointer to a struct containing members corresponding
+// to possible metadata include parameters (see http://tinyurl.com/nysdjly).
+//
+// It returns the fully qualified id of the entity.
+//
+// The name of the struct member is translated to
+// a lower case hyphen-separated form; for example,
+// ArchiveSize becomes "archive-size", and BundleMachineCount
+// becomes "bundle-machine-count", but may also
+// be specified in the field's tag
+//
+// This example will fill in the result structure with information
+// about the given id, including information on its archive
+// size (include archive-size), upload time (include archive-upload-time)
+// and digest (include extra-info/digest).
+//
+//	var result struct {
+//		ArchiveSize params.ArchiveSizeResponse
+//		ArchiveUploadTime params.ArchiveUploadTimeResponse
+//		Digest string `csclient:"extra-info/digest"`
+//	}
+//	id, err := client.Meta(id, &result)
+func (c *Client) Meta(id *charm.Reference, result interface{}) (*charm.Reference, error) {
+	if result == nil {
+		return nil, fmt.Errorf("expected valid result pointer, not nil")
+	}
+	resultv := reflect.ValueOf(result)
```

Comment:
could you use httprequest.Unmarshal to do this operation, to save having too much reflection in the system. It might change the semantics slightly though.